File: kafka-connect-aws-s3/src/test/scala/io/lenses/streamreactor/connect/aws/s3/source/SourceRecordConverterTest.scala
/*
 * Copyright 2017-2023 Lenses.io Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.lenses.streamreactor.connect.aws.s3.source

import cats.implicits.catsSyntaxOptionId
import io.lenses.streamreactor.connect.aws.s3.model.location.S3Location
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

import java.time.Instant

class SourceRecordConverterTest extends AnyFlatSpec with Matchers {

  "fromSourcePartition" should "convert S3Location to Map" in {
    SourceRecordConverter.fromSourcePartition(S3Location("test-bucket", "test-prefix".some)) should contain allOf (
      "container" -> "test-bucket",
      "prefix"    -> "test-prefix"
    )
  }

  "fromSourcePartition" should "convert S3Location without prefix to Map" in {
    SourceRecordConverter.fromSourcePartition(S3Location("test-bucket")) should contain allOf (
      "container" -> "test-bucket",
      "prefix"    -> ""
    )
  }

  "fromSourceOffset" should "convert S3Location to Map" in {
    val nowInst = Instant.now
    SourceRecordConverter.fromSourceOffset(
      S3Location("test-bucket", "test-prefix".some).withPath("test-path"),
      100L,
      nowInst,
    ) should contain allOf (
      "path" -> "test-path",
      "line" -> "100",
      "ts"   -> nowInst.toEpochMilli.toString,
    )
  }

}

File: java-connectors/kafka-connect-gcp-pubsub/src/test/java/io/lenses/streamreactor/connect/gcp/pubsub/source/mapping/SourceRecordConverterTest.java
/*
 * Copyright 2017-2024 Lenses.io Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.lenses.streamreactor.connect.gcp.pubsub.source.mapping;

import static io.lenses.streamreactor.connect.gcp.pubsub.source.mapping.MappingConfig.COMPATIBILITY_MAPPING_CONFIG;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.UnsupportedEncodingException;
import java.util.Map;

import org.apache.kafka.connect.data.Struct;
import org.junit.jupiter.api.Test;

import com.google.protobuf.ByteString;
import com.google.protobuf.Timestamp;
import com.google.pubsub.v1.PubsubMessage;

import io.lenses.streamreactor.connect.gcp.pubsub.source.subscriber.PubSubMessageData;
import io.lenses.streamreactor.connect.gcp.pubsub.source.subscriber.SourceOffset;
import io.lenses.streamreactor.connect.gcp.pubsub.source.subscriber.SourcePartition;
import lombok.val;

class SourceRecordConverterTest {

  private static final Map<String, String> ATTRIBUTE_MAP =
      Map.of(
          "key", "value"
      );

  private static final String ATTRIBUTE_MAP_STRING = "{\"key\":\"value\"}";

  @Test
  void shouldConvertDataForCompatibilityMode() throws UnsupportedEncodingException {
    val converted = new SourceRecordConverter(COMPATIBILITY_MAPPING_CONFIG).convert(setUpDataForCompatibilityMode());

    assertEquals(Map.of("message.id", "messageId1"), converted.sourceOffset());
    assertEquals(Map.of(
        "project.id", "projectId1",
        "topic.id", "topicId1",
        "subscription.id", "subscriptionId1"),
        converted.sourcePartition());
    assertEquals("targetTopicName", converted.topic());

    val connectRecord = (Struct) converted.value();
    assertEquals("My data", connectRecord.get("MessageData"));
    assertEquals(ATTRIBUTE_MAP_STRING, connectRecord.get("AttributeMap"));
  }

  private static PubSubMessageData setUpDataForCompatibilityMode() throws UnsupportedEncodingException {
    return new PubSubMessageData(
        new SourcePartition("projectId1", "topicId1", "subscriptionId1"),
        new SourceOffset("messageId1"),
        PubsubMessage.newBuilder()
            .setMessageId("messageId")
            .setData(
                ByteString.copyFrom("My data", "UTF-8")
            )
            .setPublishTime(Timestamp.newBuilder().build())
            .putAllAttributes(
                ATTRIBUTE_MAP
            ).build(),
        "targetTopicName"
    );
  }

}
